File: problems/0835-image-overlap/solution.erl
-spec largest_overlap(Img1 :: [[integer()]], Img2 :: [[integer()]]) -> integer().
largest_overlap(Img1, Img2) ->
    M1 = to_coord_list(Img1),
    M2 = to_coord_list(Img2),
    Translations = [{X2 - X1, Y2 - Y1} || {X1, Y1} <- M1, {X2, Y2} <- M2],
    count_max(Translations).


count_max([]) -> 0;
count_max(Translations) ->
    Counts = lists:foldl(fun(T, Map) ->
        maps:update_with(T, fun(V) -> V + 1 end, 1, Map)
    end, #{}, Translations),
    maps:fold(fun(_, V, Max) -> max(V, Max) end, 0, Counts).

to_coord_list(Grid) ->
    {_, Coords} = lists:foldl(fun(Row, {Y, Acc}) ->
        {_, RowCoords} = lists:foldl(fun
            (1, {X, RowAcc}) -> {X + 1, [{X, Y} | RowAcc]};
            (_, {X, RowAcc}) -> {X + 1, RowAcc}
        end, {0, []}, Row),
        {Y + 1, RowCoords ++ Acc}
    end, {0, []}, Grid),
    Coords.

File: problems/0835-image-overlap/solution.ex
defmodule Solution do
  @spec largest_overlap(img1 :: [[integer]], img2 :: [[integer]]) :: integer
  def largest_overlap(img1, img2) do
    m1 = to_coord_list(img1)
    m2 = to_coord_list(img2)
    translations = for {x1, y1} <- m1, {x2, y2} <- m2, do: {x2 - x1, y2 - y1}
    count_max(translations)
  end

  defp count_max([]), do: 0

  defp count_max(translations) do
    counts =
      Enum.reduce(translations, %{}, fn t, map ->
        Map.update(map, t, 1, &(&1 + 1))
      end)

    Enum.reduce(counts, 0, fn {_, v}, max -> max(v, max) end)
  end

  defp to_coord_list(grid) do
    grid
    |> Enum.with_index()
    |> Enum.flat_map(fn {row, y} ->
      row
      |> Enum.with_index()
      |> Enum.flat_map(fn
        {1, x} -> [{x, y}]
        {_, _} -> []
      end)
    end)
  end
end
